### DISCUSSION TITLE
Cloudwatch should escape refId when not valid

### DISCUSSION THREAD

**Discussion by rhowe** [grafana/grafana#39715] [Data Sources]
**What happened**:

I added 27 cloudwatch queries to a panel and got a metric request error.

This is because the 27th query was auto-assigned an ID of `[` which is not valid per the Cloudwatch API.

**What you expected to happen**:

27 series to be visualised

**How to reproduce it (as minimally and precisely as possible)**:

Add a cloudwatch query to a panel and set the `Id` to `[`

**Anything else we need to know?**:

I was adding these queries dynamically using grafonnet, making assigning IDs manually a bit painful.

**Environment**:
- Grafana version: 8.1.2
- Data source type & version: CloudWatch
- OS Grafana is installed on: Linux
- User OS & Browser: N/A
- Grafana plugins: N/A
- Others: N/A

**Comment by zoltanbedi** [comment]
I can see that a query with an id [ fails however I think this is not a bug in grafana. You should assign the query a proper id.

**Comment by rhowe** [comment]
> I can see that a query with an id [ fails however I think this is not a bug in grafana. You should assign the query a proper id.

I agree that I should assign an ID which is compatible with the cloudwatch API, however the ID of `[` was generated by grafana itself - it assigns IDs sequentially, starting at `A`, then `B` through to `Z`. After `Z` it then auto-assigns an ID of `[` which is where things go south with Cloudwatch.

I'm generating these queries dynamically using Grafonnet and relying on grafana's automatic ID assignment. I might be able to work around it with some jsonnet to generate IDs myself but it shouldn't be necessary. Grafana's default behaviour should not lead to failed queries IMO.

**Comment by rhowe** [comment]
This is actually also a bug in grafonnet - the `Id` of the query isn't the issue, it's the `refId` which grafonnet forcibly generates (matching Grafana behaviour) and which can generate values which are not compatible with Cloudwatch.

If Grafana is to permit query refIds which are not compatible with Cloudwatch, I think the Cloudwatch datasource should escape them in some way before passing them to the Cloudwatch API. It already does this by prepending `query` to the refId, but that's not suffficient.

**Comment by zoltanbedi** [comment]
I think we can convert this to a feature request then.

**Comment by sarahzinger** [comment]
As part of our work to move from [discussions to issues,](https://github.com/grafana/grafana/discussions/73424) we're reviewing some older discussions. It looks like this one hasn't had a lot of traction in a while so I'm going to close it for now. But if you're still running into issues around invalid refIds feel free to open an issue and we can take another look!
